File: backend/src/main/java/com/finman/service/TransactionService.java
package com.finman.service;

import com.finman.dto.request.TransactionCreateRequest;
import com.finman.dto.request.TransactionUpdateRequest;
import com.finman.dto.response.TransactionResponse;
import com.finman.dto.response.TransactionSummaryResponse;
import com.finman.entity.Account;
import com.finman.entity.Category;
import com.finman.entity.Transaction;
import com.finman.entity.User;
import com.finman.entity.enums.AccountType;
import com.finman.entity.enums.TransactionType;
import com.finman.exception.BusinessValidationException;
import com.finman.exception.ResourceNotFoundException;
import com.finman.repository.AccountRepository;
import com.finman.repository.CategoryRepository;
import com.finman.repository.TransactionRepository;
import com.finman.repository.UserRepository;
import jakarta.persistence.criteria.Predicate;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.YearMonth;
import java.util.ArrayList;
import java.util.List;

@Service
public class TransactionService {

    private final TransactionRepository transactionRepository;
    private final AccountRepository accountRepository;
    private final CategoryRepository categoryRepository;
    private final UserRepository userRepository;

    public TransactionService(
            TransactionRepository transactionRepository,
            AccountRepository accountRepository,
            CategoryRepository categoryRepository,
            UserRepository userRepository) {
        this.transactionRepository = transactionRepository;
        this.accountRepository = accountRepository;
        this.categoryRepository = categoryRepository;
        this.userRepository = userRepository;
    }

    @Transactional
    public TransactionResponse createTransaction(Long userId, TransactionCreateRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("Người dùng không tồn tại"));

        if (request.getAmount() == null || request.getAmount() <= 0) {
            throw new BusinessValidationException("Số tiền giao dịch phải lớn hơn 0");
        }

        Account account = accountRepository.findByIdAndUserId(request.getAccountId(), userId)
                .orElseThrow(() -> new ResourceNotFoundException("Tài khoản không tồn tại hoặc bạn không có quyền truy cập"));

        if (Boolean.TRUE.equals(account.getIsArchived())) {
            throw new BusinessValidationException("Không thể ghi nhận giao dịch cho tài khoản đã lưu trữ");
        }

        Category category = categoryRepository.findAccessibleCategory(request.getCategoryId(), userId)
                .orElseThrow(() -> new ResourceNotFoundException("Danh mục không tồn tại hoặc bạn không có quyền truy cập"));

        if (!request.getType().name().equals(category.getType().name())) {
            throw new BusinessValidationException("Loại giao dịch (" + request.getType() + ") không khớp với loại danh mục (" + category.getType() + ")");
        }

        // Cập nhật số dư tài khoản tương ứng
        applyBalanceImpact(account, request.getType(), request.getAmount());
        accountRepository.save(account);

        Transaction transaction = new Transaction(
                user,
                account,
                category,
                request.getType(),
                request.getAmount(),
                request.getTransactionDate(),
                request.getNote() != null ? request.getNote().trim() : null
        );

        Transaction saved = transactionRepository.save(transaction);
        return TransactionResponse.from(saved);
    }

    @Transactional
    public TransactionResponse updateTransaction(Long userId, Long transactionId, TransactionUpdateRequest request) {
        Transaction transaction = transactionRepository.findByIdAndUserId(transactionId, userId)
                .orElseThrow(() -> new ResourceNotFoundException("Giao dịch không tồn tại hoặc bạn không có quyền truy cập"));

        if (request.getAmount() == null || request.getAmount() <= 0) {
            throw new BusinessValidationException("Số tiền giao dịch phải lớn hơn 0");
        }

        Account currentAccount = transaction.getAccount();
        // 1. Hoàn tác tác động số dư của giao dịch cũ
        revertBalanceImpact(currentAccount, transaction.getType(), transaction.getAmount());

        // 2. Xác thực tài khoản mới (nếu đổi)
        Account targetAccount = currentAccount;
        if (!currentAccount.getId().equals(request.getAccountId())) {
            targetAccount = accountRepository.findByIdAndUserId(request.getAccountId(), userId)
                    .orElseThrow(() -> new ResourceNotFoundException("Tài khoản đích không tồn tại hoặc bạn không có quyền truy cập"));

            if (Boolean.TRUE.equals(targetAccount.getIsArchived())) {
                throw new BusinessValidationException("Không thể chuyển giao dịch sang tài khoản đã lưu trữ");
            }
            // Lưu lại số dư ví cũ đã hoàn tác
            accountRepository.save(currentAccount);
        }

        Category category = categoryRepository.findAccessibleCategory(request.getCategoryId(), userId)
                .orElseThrow(() -> new ResourceNotFoundException("Danh mục không tồn tại hoặc bạn không có quyền truy cập"));

        if (!request.getType().name().equals(category.getType().name())) {
            throw new BusinessValidationException("Loại giao dịch (" + request.getType() + ") không khớp với loại danh mục (" + category.getType() + ")");
        }

        // 3. Áp dụng tác động số dư mới cho tài khoản đích
        applyBalanceImpact(targetAccount, request.getType(), request.getAmount());
        accountRepository.save(targetAccount);

        // 4. Cập nhật các trường dữ liệu
        transaction.setAccount(targetAccount);
        transaction.setCategory(category);
        transaction.setType(request.getType());
        transaction.setAmount(request.getAmount());
        transaction.setTransactionDate(request.getTransactionDate());
        transaction.setNote(request.getNote() != null ? request.getNote().trim() : null);

        Transaction updated = transactionRepository.save(transaction);
        return TransactionResponse.from(updated);
    }

    @Transactional
    public void deleteTransaction(Long userId, Long transactionId) {
        Transaction transaction = transactionRepository.findByIdAndUserId(transactionId, userId)
                .orElseThrow(() -> new ResourceNotFoundException("Giao dịch không tồn tại hoặc bạn không có quyền truy cập"));

        Account account = transaction.getAccount();
        // Hoàn tác số dư của giao dịch khi bị xóa
        revertBalanceImpact(account, transaction.getType(), transaction.getAmount());
        accountRepository.save(account);

        transactionRepository.delete(transaction);
    }

    @Transactional(readOnly = true)
    public TransactionResponse getTransactionById(Long userId, Long transactionId) {
        Transaction transaction = transactionRepository.findByIdAndUserId(transactionId, userId)
                .orElseThrow(() -> new ResourceNotFoundException("Giao dịch không tồn tại hoặc bạn không có quyền truy cập"));
        return TransactionResponse.from(transaction);
    }

    @Transactional(readOnly = true)
    public Page<TransactionResponse> getTransactions(
            Long userId,
            String month,
            LocalDate startDate,
            LocalDate endDate,
            Long accountId,
            Long categoryId,
            TransactionType type,
            String search,
            Pageable pageable) {

        DateRange range = resolveDateRange(month, startDate, endDate);

        Specification<Transaction> spec = (root, query, cb) -> {
            List<Predicate> predicates = new ArrayList<>();

            // 1. Phân quyền User (Multi-tenant)
            predicates.add(cb.equal(root.get("user").get("id"), userId));

            // 2. Khoảng ngày
            if (range.startDate != null) {
                predicates.add(cb.greaterThanOrEqualTo(root.get("transactionDate"), range.startDate));
            }
            if (range.endDate != null) {
                predicates.add(cb.lessThanOrEqualTo(root.get("transactionDate"), range.endDate));
            }

            // 3. Ví / Tài khoản
            if (accountId != null) {
                predicates.add(cb.equal(root.get("account").get("id"), accountId));
            }

            // 4. Danh mục
            if (categoryId != null) {
                predicates.add(cb.equal(root.get("category").get("id"), categoryId));
            }

            // 5. Loại giao dịch (INCOME / EXPENSE)
            if (type != null) {
                predicates.add(cb.equal(root.get("type"), type));
            }

            // 6. Tìm kiếm từ khóa (trong note hoặc tên category)
            if (search != null && !search.isBlank()) {
                String pattern = "%" + search.trim().toLowerCase() + "%";
                Predicate noteLike = cb.like(cb.lower(root.get("note")), pattern);
                Predicate catLike = cb.like(cb.lower(root.get("category").get("name")), pattern);
                predicates.add(cb.or(noteLike, catLike));
            }

            return cb.and(predicates.toArray(new Predicate[0]));
        };

        // Mặc định sắp xếp theo ngày giao dịch giảm dần và ID giảm dần
        Pageable effectivePageable = pageable.getSort().isSorted()
                ? pageable
                : PageRequest.of(pageable.getPageNumber(), pageable.getPageSize(), Sort.by(Sort.Direction.DESC, "transactionDate", "id"));

        return transactionRepository.findAll(spec, effectivePageable)
                .map(TransactionResponse::from);
    }

    @Transactional(readOnly = true)
    public TransactionSummaryResponse getTransactionSummary(
            Long userId,
            String month,
            LocalDate startDate,
            LocalDate endDate,
            Long accountId) {

        DateRange range = resolveDateRange(month, startDate, endDate);
        LocalDate sDate = range.startDate != null ? range.startDate : LocalDate.of(1970, 1, 1);
        LocalDate eDate = range.endDate != null ? range.endDate : LocalDate.of(2099, 12, 31);

        Long totalIncome;
        Long totalExpense;
        Long totalCount;

        if (accountId != null) {
            totalIncome = transactionRepository.sumAmountByUserIdAndAccountIdAndTypeAndDateBetween(
                    userId, accountId, TransactionType.INCOME, sDate, eDate);
            totalExpense = transactionRepository.sumAmountByUserIdAndAccountIdAndTypeAndDateBetween(
                    userId, accountId, TransactionType.EXPENSE, sDate, eDate);
            totalCount = transactionRepository.countByUserIdAndDateBetween(userId, sDate, eDate);
        } else {
            totalIncome = transactionRepository.sumAmountByUserIdAndTypeAndDateBetween(
                    userId, TransactionType.INCOME, sDate, eDate);
            totalExpense = transactionRepository.sumAmountByUserIdAndTypeAndDateBetween(
                    userId, TransactionType.EXPENSE, sDate, eDate);
            totalCount = transactionRepository.countByUserIdAndDateBetween(userId, sDate, eDate);
        }

        long netFlow = (totalIncome != null ? totalIncome : 0L) - (totalExpense != null ? totalExpense : 0L);
        return new TransactionSummaryResponse(totalIncome, totalExpense, netFlow, totalCount);
    }

    /**
     * Áp dụng thay đổi số dư theo từng loại tài khoản.
     */
    private void applyBalanceImpact(Account account, TransactionType type, Long amount) {
        if (account.getType() == AccountType.CREDIT_CARD) {
            // Với thẻ tín dụng, currentBalance đại diện cho dư nợ
            if (type == TransactionType.EXPENSE) {
                account.setCurrentBalance(account.getCurrentBalance() + amount);
            } else if (type == TransactionType.INCOME) {
                account.setCurrentBalance(account.getCurrentBalance() - amount);
            }
        } else {
            // Với ví tiền mặt và ngân hàng, currentBalance là tài sản khả dụng
            if (type == TransactionType.INCOME) {
                account.setCurrentBalance(account.getCurrentBalance() + amount);
            } else if (type == TransactionType.EXPENSE) {
                account.setCurrentBalance(account.getCurrentBalance() - amount);
            }
        }
    }

    /**
     * Hoàn tác số dư giao dịch cũ theo từng loại tài khoản.
     */
    private void revertBalanceImpact(Account account, TransactionType type, Long amount) {
        if (account.getType() == AccountType.CREDIT_CARD) {
            if (type == TransactionType.EXPENSE) {
                account.setCurrentBalance(account.getCurrentBalance() - amount);
            } else if (type == TransactionType.INCOME) {
                account.setCurrentBalance(account.getCurrentBalance() + amount);
            }
        } else {
            if (type == TransactionType.INCOME) {
                account.setCurrentBalance(account.getCurrentBalance() - amount);
            } else if (type == TransactionType.EXPENSE) {
                account.setCurrentBalance(account.getCurrentBalance() + amount);
            }
        }
    }

    /**
     * Xác định khoảng ngày bắt đầu và kết thúc từ tham số month hoặc startDate/endDate.
     */
    private DateRange resolveDateRange(String month, LocalDate startDate, LocalDate endDate) {
        LocalDate resolvedStart = startDate;
        LocalDate resolvedEnd = endDate;

        if (month != null && !month.isBlank()) {
            try {
                YearMonth ym = YearMonth.parse(month.trim());
                if (resolvedStart == null) {
                    resolvedStart = ym.atDay(1);
                }
                if (resolvedEnd == null) {
                    resolvedEnd = ym.atEndOfMonth();
                }
            } catch (Exception ignored) {
            }
        }

        return new DateRange(resolvedStart, resolvedEnd);
    }

    private static class DateRange {
        final LocalDate startDate;
        final LocalDate endDate;

        DateRange(LocalDate startDate, LocalDate endDate) {
            this.startDate = startDate;
            this.endDate = endDate;
        }
    }
}

File: frontend/src/services/transactionService.ts
import { api } from './api';
import type { Transaction, TransactionType } from '../types';

interface ApiResponse<T> {
  success: boolean;
  message: string;
  data: T;
}

interface SpringPage<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
}

export interface TransactionSummary {
  totalIncome: number;
  totalExpense: number;
  netCashFlow: number;
  totalTransactions: number;
}

export interface TransactionCreatePayload {
  accountId: number;
  categoryId: number;
  type: TransactionType;
  amount: number;
  transactionDate: string; // YYYY-MM-DD
  note?: string;
}

export interface TransactionUpdatePayload {
  accountId?: number;
  categoryId?: number;
  type?: TransactionType;
  amount?: number;
  transactionDate?: string;
  note?: string;
}

export interface TransactionFilterParams {
  month?: string; // YYYY-MM
  startDate?: string; // YYYY-MM-DD
  endDate?: string; // YYYY-MM-DD
  accountId?: number;
  categoryId?: number;
  type?: TransactionType;
  search?: string;
  page?: number;
  size?: number;
}

export const transactionService = {
  /**
   * Lấy danh sách giao dịch từ cơ sở dữ liệu backend
   */
  async getTransactions(params?: TransactionFilterParams): Promise<Transaction[]> {
    const res = await api.get<ApiResponse<SpringPage<any> | any[]>>('/transactions', {
      params: {
        ...params,
        size: params?.size || 100, // Get up to 100 recent transactions
      },
    });

    const rawData = res.data.data;
    const rawList: any[] = Array.isArray(rawData) ? rawData : (rawData?.content || []);

    return rawList.map((item) => ({
      id: item.id,
      amount: item.amount,
      type: item.type,
      date: item.transactionDate || item.date,
      time: item.createdAt ? new Date(item.createdAt).toLocaleTimeString('vi-VN', { hour: '2-digit', minute: '2-digit' }) : undefined,
      note: item.note,
      createdAt: item.createdAt,
      account: {
        id: item.account?.id,
        name: item.account?.name,
        type: item.account?.type,
        accountNumber: item.account?.accountNumber,
        currentBalance: 0,
      },
      category: {
        id: item.category?.id,
        name: item.category?.name,
        type: item.category?.type,
        icon: item.category?.icon || 'payments',
        color: item.category?.color,
      },
    }));
  },

  /**
   * Lấy thống kê dòng tiền tổng thể từ cơ sở dữ liệu
   */
  async getTransactionSummary(params?: {
    month?: string;
    startDate?: string;
    endDate?: string;
    accountId?: number;
  }): Promise<TransactionSummary> {
    const res = await api.get<ApiResponse<TransactionSummary>>('/transactions/summary', { params });
    return res.data.data;
  },

  /**
   * Tạo giao dịch mới trong database và tự động điều chỉnh số dư tài khoản
   */
  async createTransaction(payload: TransactionCreatePayload): Promise<Transaction> {
    const res = await api.post<ApiResponse<any>>('/transactions', payload);
    const item = res.data.data;
    return {
      id: item.id,
      amount: item.amount,
      type: item.type,
      date: item.transactionDate || item.date,
      time: item.createdAt ? new Date(item.createdAt).toLocaleTimeString('vi-VN', { hour: '2-digit', minute: '2-digit' }) : undefined,
      note: item.note,
      createdAt: item.createdAt,
      account: {
        id: item.account?.id,
        name: item.account?.name,
        type: item.account?.type,
        accountNumber: item.account?.accountNumber,
        currentBalance: 0,
      },
      category: {
        id: item.category?.id,
        name: item.category?.name,
        type: item.category?.type,
        icon: item.category?.icon || 'payments',
        color: item.category?.color,
      },
    };
  },

  /**
   * Cập nhật giao dịch và hoàn tác/áp dụng số dư tương ứng
   */
  async updateTransaction(id: number, payload: TransactionUpdatePayload): Promise<Transaction> {
    const res = await api.put<ApiResponse<any>>(`/transactions/${id}`, payload);
    const item = res.data.data;
    return {
      id: item.id,
      amount: item.amount,
      type: item.type,
      date: item.transactionDate || item.date,
      time: item.createdAt ? new Date(item.createdAt).toLocaleTimeString('vi-VN', { hour: '2-digit', minute: '2-digit' }) : undefined,
      note: item.note,
      createdAt: item.createdAt,
      account: {
        id: item.account?.id,
        name: item.account?.name,
        type: item.account?.type,
        accountNumber: item.account?.accountNumber,
        currentBalance: 0,
      },
      category: {
        id: item.category?.id,
        name: item.category?.name,
        type: item.category?.type,
        icon: item.category?.icon || 'payments',
        color: item.category?.color,
      },
    };
  },

  /**
   * Xóa giao dịch khỏi cơ sở dữ liệu và hoàn tác số dư ví
   */
  async deleteTransaction(id: number): Promise<void> {
    await api.delete<ApiResponse<void>>(`/transactions/${id}`);
  },
};
